controller/array_action/array_mediator_ds8k.py

```diff
@@ -0,0 +1,412 @@
+FC_STATUS = 'state'
+FC_IS_PERSISTENT = 'persistent'  # enabled|disabled
+FC_IS_RECORDING = 'recording'  # enabled|multinc|disabled|None
+FC_IS_BGCOPY = 'backgroundcopy'  # enabled|disabled
+FC_ID = 'id'
+
+# pprc
+PPRC_SOURCE_VOLUME_ID = 'source_volume'
+PPRC_TARGET_VOLUME_ID = 'target_volume'
+# PPRC_SOURCE_ESS_NAME = 'system'
+PPRC_SOURCE_SYSTEM_ID = 'source_system'
+PPRC_TARGET_SYSTEM_ID = 'target_system'
+PPRC_STATUS = 'state'
+PPRC_REMOTE_COPY_TYPE = 'type'  # metromirror|globalcopy|unknown
+
+# ioport
+IOPORT_NAME = 'id'
+IOPORT_WWPN = 'wwpn'
+# IOPORT_WWNN = ''
+IOPORT_PORT_SPEED = 'speed'
+IOPORT_STATUS = 'state'
+IOPORT_STATUS_ONLINE = 'online'
+IOPORT_LOCATION = 'loc'
+IOPORT_ENCLOSURE_NUMBER = 'io_enclosure'
+
+# host
+HOST_ID = 'name'
+HOST_NAME = 'name'
+HOST_STATE = 'state'
+HOST_TYPE = 'hosttype'  # VMWare
+HOST_VOLUME_MAPPINGS = 'mappings_briefs'
+HOST_VOLUME_MAPPING_VOLUME_ID = 'volume_id'
+HOST_VOLUME_MAPPING_LUN_ID = 'lunid'
+HOST_PORTS = 'host_ports_briefs'
+HOST_ADDRESS_DISCOVERY = 'addrdiscovery'
+
+# host port
+HOST_PORT_WWPN = 'wwpn'
+HOST_PORT_HOST_ID = 'host'
+HOST_PORT_IOPORTS = 'login_ports'
+HOST_PORT_STATE = 'state'
+HOST_PORT_TYPE = 'hosttype'
+
+# host mapping
+HOST_MAPPING_LUN_ID = 'lunid'
+HOST_MAPPING_VOLUME_ID = 'volume'
+
+# user
+USER_NAME = 'name'
+USER_STATE = 'state'
+USER_GROUP = 'group'
+
+# lss
+LSS_ID = 'id'
+LSS_TYPE = 'type'
+LSS_GROUP = 'group'
+
+# io enclosure
+IOENCLOSURE_ID = 'id'
+IOENCLOSURE_NAME = 'name'
+IOENCLOSURE_STATE = 'state'
+
+# node
+NODE_ID = 'id'
+NODE_STATUS = 'state'
+
+# marray
+MARRAY_ID = 'id'
+MARRAY_DISK_CLASS = 'disk_class'
+MARRAY_POOL_ID = 'pool'
+
+# response keys
+RES_STATUS = 'status'
+RES_CODE = 'code'
+RES_MSG = 'message'
+
+# response values
+RES_SUCCESS_STAUTS = ('ok', 'successful', )
+
+
+def parse_version(bundle):
+    """
+    Parse ds8k version number form bundle.
+
+    rule is:
+    87.51.34.0 => 7.5.1
+    87.50.21.1 => 7.5.0
+    88.0.151.0 => 8.0.0
+    """
+
+    v1, v2, _ = bundle.split('.', 2)
+    v2 = '{0:0{1}}'.format(int(v2), 2)
+
+    return '.'.join([v1[-1], v2[0], v2[1]])
+
+
+def get_volume_id_from_scsi_identifier(scsi_id):
+    return scsi_id[-4:]
+
+
+def build_kwargs_from_capabilities(capabilities, pool_id, name, size):
+    cli_kwargs = {}
+    size_in_gib = int(math.ceil(convert_size_bytes_to_gib(size)))
+    cli_kwargs.update({
+        'pool_id': pool_id,
+        'capacity_in_gib': size_in_gib,
+        'tp': 'none',
+        'volume_names_list': [shorten_volume_name(name), ],
+    })
+    capability = capabilities.get(config.CAPABILITIES_SPACEEFFICIENCY)
+    if capability:
+        capability = capability.lower()
+        if capability == config.CAPABILITY_THIN:
+            cli_kwargs.update({'tp': 'ese'})
+
+    return cli_kwargs
+
+
+# shorten volume name to 16 characters if it is too long.
+def shorten_volume_name(name):
+    if len(name) <= 16:
+        return name
+    return sha1(name.encode()).hexdigest()[-16:]
```

Comment:
which ones? then at least use base58. base16 is very limited...

Reply:
You are using base16, by calling `hexdigest` (hex is short for hexadecimal, which means 16). This is the part that I meant that should be replaced with base58.

I agree that a base conversion alone is not enough for hashing purposes.

Reply:
[hexdigest](https://docs.python.org/3/library/hashlib.html#hashlib.hash.hexdigest) is like the binary `digest`, but contains "only hexadecimal digits". i.e. it encodes the binary hash with only [16](https://en.m.wikipedia.org/?title=Base16&redirect=no) different characters, not utilizing most of the alphabet letters.
And since the length is very limited, it unnecessarily increases the chance of a hash collision, comparing to base58.

Reply:
Let's stop here, I prefer sha1().hexdigest()

Reply:
We should not overlook this.
I don't see why not use base58 - It has a much better [space efficiency](https://en.wikipedia.org/wiki/Binary-to-text_encoding#Encoding_standards).
`sha1().hexdigest()[-16:]` will use only 8 bytes of the hash (e.g. 'ffffffffffffffff').
`base58.b58encode(sha1().digest()).decode()[-16:]` will use 11 bytes! (e.g. '26Uw2Vvq8EnJ7hRG')
The more bytes you use from the hash, the lower the chance of a hash collision.
It is not a matter of style, it has a functional impact.

*Note that if ds8k volume name is case insensitive, use base36 instead

Reply:
ok, I accepted.

Reply:
Thanks.
Just to make sure:
you wrote "the id field of all resources are case insensitive", is volume name also case insensitive?

Reply:
writing here again, in case you missed this

Reply:
no